### PR TITLE
Bump base image Ubuntu to 24.04, Bump gdal to 3.9

### PR DESCRIPTION
**Pull request by conbrad**
- Rebuilt `wps-api-base`to use `osgeo/gdal:ubuntu-small-3.9.2` which comes with Ubuntu 24.04, python 3.12.3 and gdal 3.9.2
- Updates tippecanoe to 2.62.5
- Installing gdal for builds requires a few more flags otherwise it won't recognize the `numpy` installation
- Images already have GID 1000 and UID 1000 so a new user with ID 1010 is created and added to existing GID 1000 for mount permissions
- Uses `cutlineWkt` instead of `cutlineSQL` from `gdal` to avoid issues with gdal postgres search paths
- A bunch of fixes to C-Haines code that started failing tests with new poetry updates
- Documentation and setup script updates to reference the new python version
- Removes `shapely` from `tileserv` project since it doesn't build with python 3.12.3 and we don't use it anymore

# Test Links:
[Landing Page](https://wps-pr-3995-e1e498-dev.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-3995-e1e498-dev.apps.silver.devops.gov.bc.ca/morecast)
[Percentile Calculator](https://wps-pr-3995-e1e498-dev.apps.silver.devops.gov.bc.ca/percentile-calculator)
[C-Haines](https://wps-pr-3995-e1e498-dev.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-3995-e1e498-dev.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-3995-e1e498-dev.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-3995-e1e498-dev.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-3995-e1e498-dev.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: .github/workflows/deployment.yml [hunk 1]
```diff
@@ -233,16 +258,21 @@ jobs:
...
+      - name: Install OpenShift CLI tools
+        uses: redhat-actions/openshift-tools-installer@v1
+        with:
+          oc: "4.14"
+
       - name: NATS Message Queue
         shell: bash
         run: |
           oc login "${{ secrets.OPENSHIFT_CLUSTER }}" --token="${{ secrets.OC4_DEV_TOKEN }}"
-          PROJ_DEV="e1e498-dev" MEMORY_REQUEST=250Mi MEMORY_LIMIT=500Mi bash openshift/scripts/oc_provision_nats.sh ${SUFFIX} apply
```
Comment (dgboss) [inline .github/workflows/deployment.yml:240]: Should we restore the memory settings?

Comment (conbrad) [reply .github/workflows/deployment.yml:240]: Yes, will do that before merging

**Comment by dgboss** [comment]
In `deployment.yaml` there are two jobs that include a step to install the oc cli but the jobs still run on ubuntu-22.04:
- deploy-dev
- deploy-dev-queue
Is this intentional?

**Comment by conbrad** [comment]
> ubuntu-22.04

Unintentional, I just can't read yaml. Thanks for catching.

File: .github/workflows/post_merge_integration.yml [hunk 2]
```diff
@@ -76,11 +76,11 @@ jobs:
 
   test-api:
     name: Python - Test with coverage
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-24.04
     strategy:
       matrix:
         # Match versions for python + R to Ubuntu 22.04 LTS at the time of writing.
```
Comment (dgboss) [inline .github/workflows/post_merge_integration.yml:82]: nit: Update comment to 'Ubuntu 24.04'.

**Comment by dgboss** [comment]
> > ubuntu-22.04
> 
> Unintentional, I just can't read yaml. Thanks for catching.

One more in that same file:
- configure-nats-server-name

File: .github/workflows/cleanup.yml [hunk 3]
```diff
@@ -9,7 +9,7 @@ on:
 jobs:
   cleanup-non-db:
     name: Delete non-DB resources associated with pr
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-24.04
     steps:
       - name: Set Variables
         shell: bash
```
Comment (dgboss) [inline .github/workflows/cleanup.yml:15]: Might as well bump up the Ubuntu version on the `cleanup-db` job in this file as well.

File: Dockerfile [hunk 4]
```diff
@@ -7,8 +7,8 @@ FROM ${DOCKER_IMAGE} AS builder
 
 # We don't want to run our app as root, so we define a worker user.
 ARG USERNAME=worker
-ARG USER_UID=1000
-ARG USER_GID=$USER_UID
+ARG USER_UID=1010
```
Comment (dgboss) [inline Dockerfile:10]: Is this necessary because the base image already has a `USERID` of 1000?

Comment (conbrad) [reply Dockerfile:10]: Yes

File: api/.python-version [hunk 5]
```diff
@@ -0,0 +1 @@
+3.12.3
```
Comment (dgboss) [inline api/.python-version:1]: Did you mean to commit this file?

File: api/app/auto_spatial_advisory/elevation.py [hunk 6]
```diff
@@ -262,12 +263,17 @@ async def process_tpi_by_firezone(run_type: RunType, run_date: date, for_date: d
...
     async with get_async_write_session_scope() as session:
-        stmt = text("SELECT id, source_identifier FROM advisory_shapes;")
+        stmt = text("SELECT id, source_identifier FROM public.advisory_shapes;")
         result = await session.execute(stmt)
 
         for row in result:
             output_path = f"/vsimem/firezone_{row[1]}.tif"
-            warp_options = gdal.WarpOptions(format="GTiff", cutlineDSName=DB_READ_STRING, cutlineSQL=f"SELECT geom FROM advisory_shapes WHERE id={row[0]}", cropToCutline=True)
+
+            input_srs = osr.SpatialReference()
```
Comment (dgboss) [inline api/app/auto_spatial_advisory/elevation.py:272]: You might be able to simplify this to:
`input_srs = hfi_masked_tpi.GetSpatialRef()`

File: api/.python-version (hunk 5 above)
Comment (conbrad) [reply api/.python-version:1]: Poetry created it I believe.

File: .github/workflows/post_merge_integration.yml (hunk 2 above)
Comment (conbrad) [reply .github/workflows/post_merge_integration.yml]: Done in: https://github.com/bcgov/wps/pull/3995/commits/095b46bf148446caf78f0c9c0eca7bb747d0e5c7

File: .github/workflows/cleanup.yml (hunk 3 above)
Comment (conbrad) [reply .github/workflows/cleanup.yml:15]: Done in: https://github.com/bcgov/wps/pull/3995/commits/dfaba153d40a319d5f8f57cd8c2df4b83ea055ef

File: api/app/auto_spatial_advisory/elevation.py (hunk 6 above)
Comment (conbrad) [reply api/app/auto_spatial_advisory/elevation.py]: Done in: https://github.com/bcgov/wps/pull/3995/commits/8ab763f4855d688383d29cd8c6646904478daf5e

**Review by dgboss — APPROVED** [review]
Great work!

File: api/.python-version (hunk 5 above)
Comment (dgboss) [reply api/.python-version:1]: I think it is a pyenv thing. Probably not a bad idea as it should ensure that devs are running Python 3.12.X locally.

File: api/app/auto_spatial_advisory/elevation.py [hunk 7]
```diff
@@ -315,7 +319,12 @@ def intersect_raster_by_firezone(threshold: int, advisory_shape_id: int, source_
     :param temp_dir: A temporary location for storing intermediate files
     """
     output_path = os.path.join(temp_dir, f"firezone_{source_identifier}_threshold_{threshold}.tif")
-    warp_options = gdal.WarpOptions(format="GTiff", cutlineDSName=DB_READ_STRING, cutlineSQL=f"SELECT geom FROM advisory_shapes WHERE id={advisory_shape_id}", cropToCutline=True)
+    warp_options = gdal.WarpOptions(
+        format="GTiff",
+        cutlineDSName=DB_READ_STRING,
+        cutlineSQL=f"SELECT geom FROM public.advisory_shapes WHERE id={advisory_shape_id}",
```
Comment (brettedw) [inline api/app/auto_spatial_advisory/elevation.py:325]: This still works?

**Review by brettedw — APPROVED** [review]
Great work! And great perseverance!

File: api/app/auto_spatial_advisory/elevation.py (hunk 7 above)
Comment (conbrad) [reply api/app/auto_spatial_advisory/elevation.py:325]: Oh this isn't used anymore (it's from the old elevation stats job). There's a ticket to clean that up: https://github.com/bcgov/wps/issues/3997

**Pull request merged**